python/lsst/dax/imgserv/getimage/imageget.py

```diff
@@ -0,0 +1,625 @@
+        cutout = self._cutout_by_data_id(data_id, ra, dec, width, height,
+                                         unit)
+        return cutout
+
+    def cutout_from_data_id_by_tract(self, tract, patch_x, patch_y, filt,
+                                     ra, dec, width, height,
+                                     unit):
+        """Returns cutout image from data id (tract, patch<x,y>, filt)
+        of specified center.
+
+        Parameters
+        ----------
+        tract : `int`
+        patch_x : `int`
+        pactch_y : `int`
+        filt : `str`
+        ra: `float`
+        dec : `float`
+            in degrees.
+        width : `float`
+        height: `float`
+        unit : `str`
+
+        Returns
+        -------
+        cutout: `afwImage.Exposure`
+
+        """
+        patch = ",".join((str(patch_x), str(patch_y)))
+        data_id = {"tract": tract, "patch": patch, "filter": filt}
+        cutout = self._cutout_by_data_id(data_id, ra, dec, width, height, unit)
+        return cutout
+
+    def cutout_from_ccd_exp_id(self, ccd_exp_id, ra, dec, width, height,
+                               unit):
+        """
+        Parameters
+        ----------
+        ccd_exp_id : `int`
+        ra : `float`
+        dec: `float`
+            in degrees.
+        width : `float`
+        height : `float`
+        unit : `str`
+            [ `pixel', 'pixels', 'arcsec' ]
+
+        Returns
+        -------
+        cutout: `afwImage.Exposure`
+
+        """
+        # Get the corresponding image(data) id from the butler
+        data_id = self.data_id_from_ccd_exp_id(ccd_exp_id)
+        if data_id:
+            # make id compatible with qResult type via custom wrapping
+            cutout = self._cutout_by_data_id(data_id, ra, dec, width, height,
+                                             unit)
+            return cutout
+
+    def cutout_from_skymap_id(self, skymap_id, filt, ra, dec, width, height,
+                              unit):
+        """
+        Parameters
+        ----------
+        skymap_id : `str`
+        filt : filter
+        ra : `float`
+        dec : `float`
+        width : `float`
+        height : `float`
+        unit : `str`
+            [ 'px', 'pix', 'pixel', 'pixels', 'arcsec']
+
+        Returns
+        -------
+        cutout: `afwImage.Exposure`
+
+        """
+        skymap = SkymapImage(self._butler, skymap_id, self._log)
+        center_coord = afwGeom.SpherePoint(ra, dec, afwGeom.degrees)
+        cutout = skymap.get(center_coord, width, height, filt, unit)
+        return cutout
+
+    def data_id_from_ccd_exp_id(self, ccd_exp_id):
+        """ The ids match the ids in _butler_keys and valid is false
+        if at least one of the ids is missing.
+
+        Parameters
+        ----------
+        ccd_exp_id: `int`
+
+        Returns
+        -------
+        data_id : `dict`
+            the list of ids derived from scienceId.
+
+        """
```

Comment:
This logic and math is pretty confusing and obtuse to me.  I have no idea what is going on here.

Reply:
This math is at the core of how LSST composes/de-composes data ids from the underlying identifying information, as implemented in our Butler.  KT has been the source of truth for this arithmetic logic.